lib/presentation/widget/shared_space_details/add_shared_space_member/add_shared_space_member_viewmodel.dart

```diff
@@ -0,0 +1,111 @@
+import 'package:linshare_flutter_app/presentation/redux/actions/shared_space_action.dart';
+import 'package:linshare_flutter_app/presentation/redux/online_thunk_action.dart';
+import 'package:linshare_flutter_app/presentation/redux/states/app_state.dart';
+import 'package:linshare_flutter_app/presentation/util/router/app_navigation.dart';
+import 'package:linshare_flutter_app/presentation/widget/base/base_viewmodel.dart';
+import 'package:redux/redux.dart';
+
+class AddSharedSpaceMemberViewModel extends BaseViewModel {
+  final AppNavigation _appNavigation;
+  final GetAutoCompleteSharingInteractor _getAutoCompleteSharingInteractor;
+  final AddSharedSpaceMemberInteractor _addSharedSpaceMemberInteractor;
+  final GetAllSharedSpaceRolesInteractor _getAllSharedSpaceRolesInteractor;
+  final GetAllSharedSpaceMembersInteractor _getAllSharedSpaceMembersInteractor;
+
+  SharedSpaceRoleName selectedRole = SharedSpaceRoleName.READER;
+
+  AddSharedSpaceMemberViewModel(
+      Store<AppState> store,
+      this._appNavigation,
+      this._getAutoCompleteSharingInteractor,
+      this._addSharedSpaceMemberInteractor,
+      this._getAllSharedSpaceRolesInteractor,
+      this._getAllSharedSpaceMembersInteractor)
+      : super(store);
+
+  void backToSharedSpacesDetails() {
+    _appNavigation.popBack();
+  }
+
+  void selectRole(SharedSpaceRoleName role) {
+    selectedRole = role;
+    _appNavigation.popBack();
+  }
+
+  Future<List<SharedSpaceMember>> getSharedSpaceMembers(SharedSpaceId sharedSpaceId) async {
+    return (await _getAllSharedSpaceMembersInteractor.execute(sharedSpaceId))
+        .map((result) => result is SharedSpaceMembersViewState ? result.members : null)
+        .getOrElse(() => []);
+  }
+
+  Future<List<AutoCompleteResult>> getAutoCompleteSharing(
+      String pattern, SharedSpaceNodeNested sharedSpace, List<SharedSpaceMember> members) async {
+    return await _getAutoCompleteSharingInteractor
+        .execute(AutoCompletePattern(pattern), AutoCompleteType.THREAD_MEMBERS,
+            threadId: ThreadId(sharedSpace.sharedSpaceId.uuid))
+        .then(
+          (viewState) => viewState.fold(
+            (failure) => <AutoCompleteResult>[],
+            (success) => (success as AutoCompleteViewState)
+                .results
+                .where((user) => !members
+                    .map((member) => member.account.mail)
+                    .contains(user.getSuggestionMail()))
+                .toList(),
+          ),
+        );
+  }
+
+  void addSharedSpaceMember(
+      SharedSpaceId sharedSpaceId, SharedSpaceMemberAutoCompleteResult autoCompleteResult) {
+    store.dispatch(
+        _addSharedSpaceAction(sharedSpaceId, AccountId(autoCompleteResult.userUuid.uuid)));
+  }
+
+  OnlineThunkAction _addSharedSpaceAction(SharedSpaceId sharedSpaceId, AccountId accountId) {
+    return OnlineThunkAction((Store<AppState> store) async {
+      final roles = (await _getAllSharedSpaceRolesInteractor.execute())
```

Comment:
I don't think it is a good idea, `role name` and `role id` are not changed too much, why we not fetch it only 1 time when open the add member screen instead of fetch it every time add member?